File: 백준/Silver/14620. 꽃길/꽃길.cc
#include<bits/stdc++.h>
using namespace std;
const int dy[]={-1,0,1,0};
const int dx[]={0,1,0,-1};
int n,ret=INT_MAX;
int a[12][12];
int visited[12][12];
bool check(int y,int x)
{
	if(visited[y][x]) return false;
	for(int i=0;i<4;i++)
	{
		int ny=y+dy[i];
		int nx=x+dx[i];
		if(ny<0 || ny>=n || nx<0 || nx>=n) return false;
		if(visited[ny][nx]) return false;
	}
	return true;
}
void go(int cnt,int sum)
{
	if(cnt==3)
	{
		ret=min(ret,sum);
		return;
	}
	for(int i=1;i<n-1;i++)
	{
		for(int j=1;j<n-1;j++)
		{
			if(check(i,j))
			{
				visited[i][j]=1;
				int cost=a[i][j];
				for(int k=0;k<4;k++)
				{
					int ny=i+dy[k];
					int nx=j+dx[k];
					visited[ny][nx]=1;
					cost+=a[ny][nx];
				}
				
				go(cnt+1,sum+cost);
				
				visited[i][j]=0;
				for(int k=0;k<4;k++)
				{
					int ny = i+dy[k];
					int nx = j+dx[k];
					visited[ny][nx]=0;
				}	
			}
		}
	}
}
int main()
{
	cin>>n;
	for(int i=0;i<n;i++)
	{
		for(int j=0;j<n;j++)
		{
			cin>>a[i][j];
		}
	}
	go(0,0);
	cout<<ret;
}

File: 백준/Silver/14620. 꽃길/꽃길.py
dy=[-1,0,1,0]
dx=[0,1,0,-1]
ret=float('inf')
def canPlace(y,x):
    if visited[y][x]: #이미 방문한 정점이면 False
        return False
    for i in range(4):
        ny=y+dy[i]
        nx=x+dx[i]
        if ny<0 or ny>=n or nx<0 or nx>=n or visited[ny][nx]: # 범위에서 벗어나면 False
            return False
    return True

def plantFlower(y,x):
    visited[y][x]=True
    c=a[y][x]
    for i in range(4):
        ny=y+dy[i]
        nx=x+dx[i]
        visited[ny][nx]=True
        c+=a[ny][nx]
    return c

def removeFlower(y,x):
    visited[y][x]=False
    for i in range(4):
        ny=y+dy[i]
        nx=x+dx[i]
        visited[ny][nx]=False

def dfs(cnt,sum):
    global ret
    if cnt==3:
        ret=min(ret,sum)
        return

    for i in range(1,n):
        for j in range(1,n):
            if(canPlace(i,j)):
                cost=plantFlower(i,j)
                dfs(cnt+1,sum+cost)
                removeFlower(i,j)

n=int(input().strip())
a=[[0] * n for _ in range(n)]
visited=[[False] * n for _ in range(n)]

for i in range(n):
    num=list(map(int,input().split()))
    for j in range(n):
        a[i][j]=num[j]

dfs(0,0) # cnt,sum
print(ret)